src/jdk.compiler/share/classes/com/sun/tools/javac/comp/TransPatterns.java

```diff
@@ -640,6 +675,167 @@ JCNewClass makeNewClass(Type ctype, List<JCExpression> args) {
+                                                  List.of(continueSwitch),
+                                                  null));
+                    }
+                    JCSwitch newSwitch = make.Switch(commonNestedExpression, nestedCases.toList());
+                    newSwitch.patternSwitch = true;
+                    JCPatternCaseLabel leadingTest =
+                            (JCPatternCaseLabel) accummulator.first().labels.head;
+                    leadingTest.guard = null;
+                    result.add(make.Case(CaseKind.STATEMENT,
+                                         List.of(leadingTest),
+                                         List.of(newSwitch),
+                                         null));
+                } else {
+                    result.addAll(accummulator);
+                }
+                accummulator.clear();
+        };
+
+        VarSymbol commonBinding = null;
+        JCExpression commonNestedExpression = null;
+        VarSymbol commonNestedBinding = null;
+
+        for (List<JCCase> c = inputCases; c.nonEmpty(); c = c.tail) {
+            VarSymbol currentBinding;
+            JCExpression currentNestedExpression;
+            VarSymbol currentNestedBinding;
+
+            if (c.head.labels.size() == 1 &&
+                c.head.labels.head instanceof JCPatternCaseLabel patternLabel &&
+                patternLabel.guard instanceof JCBinary binOp &&
+                binOp.lhs instanceof JCInstanceOf instanceofCheck &&
+                instanceofCheck.pattern instanceof JCBindingPattern binding) {
+                currentBinding = ((JCBindingPattern) patternLabel.pat).var.sym;
+                currentNestedExpression = instanceofCheck.expr;
+                currentNestedBinding = binding.var.sym;
+            } else if (c.head.labels.size() == 1 &&
+                c.head.labels.head instanceof JCPatternCaseLabel patternLabel &&
+                patternLabel.guard instanceof JCInstanceOf instanceofCheck &&
+                instanceofCheck.pattern instanceof JCBindingPattern binding) {
+                currentBinding = ((JCBindingPattern) patternLabel.pat).var.sym;
+                currentNestedExpression = instanceofCheck.expr;
+                currentNestedBinding = binding.var.sym;
+            } else {
+                currentBinding = null;
+                currentNestedExpression = null;
+                currentNestedBinding = null;
+            }
+            if (commonBinding == null) {
+                if (currentBinding != null) {
+                    commonBinding = currentBinding;
+                    commonNestedExpression = currentNestedExpression;
+                    commonNestedBinding = currentNestedBinding;
+                    accummulator.add(c.head);
+                } else {
+                    result.add(c.head);
+                }
+            } else if (currentBinding != null &&
+                       commonBinding.type.tsym == currentBinding.type.tsym &&
```

Comment:
question, what about having a map relating commonBindings and accumulators, shouldn't this allow you to group common patterns even if they are not consecutive?

Reply:
This would mean we would change the order of the cases, which is something I'd like to avoid in this patch, for several reasons. Partly because there's a fairly big set of cases where it is unclear if we can do the reordering (and the exact rules when we can do reordering are debatable, and sometimes depend on very particular aspects of the spec), and partly because this patch seems complex enough without reordering.

One possibility we could do case reordering as a separate step before coalescing the prefixes, as simple use of map would not necessarily allow to avoid coalescing in cases like:
```
switch (o) {
     case I1 i when <condition1> -> {}
     case I2 i when <condition2> -> {}
     case I1 i when <condition3> -> {}
}
```
(where `I1` and `I2` are unrelated interfaces, we cannot easily change the order of the cases here, as `o` may be an object that implements both, and `<condition1>` might be `false`).

Reply:
got it I agree, thanks